contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,173 @@
+pragma solidity ^0.8.0;
+
+import "../interfaces/LinkTokenInterface.sol";
+import "../interfaces/BlockHashStoreInterface.sol";
+
+import "../vendor/Ownable.sol";
+
+import "./VRF.sol";
+
+contract VRFCoordinatorV2 is VRF, Ownable {
+
+    LinkTokenInterface internal LINK;
+    BlockHashStoreInterface internal blockHashStore;
+
+    uint256 currentSubId;
+    struct Subscription {
+        uint256 subId;
+        address owner; // Owner can fund/withdraw/cancel the sub
+        address[] consumers; // List of addresses which can consume using this subscription.
+        uint256 balance; // Common balance used for all consumer requests.
+    }
+    mapping(uint256 /* subId */ => Subscription /* subscription */) private subscriptions;
+
+    struct Oracle {
+        address oracle;
+        bytes32 keyHash;
+        mapping(address => uint256) nonces;
+    }
+
+    mapping(bytes32 /* keyHash */ => Oracle) oracles;
+
+    struct Callback {
+        address callbackContract; // Requesting contract, which will receive response
+        bytes32 requestBlockNum; // Request block number
+        uint256 numWords;
+        uint256 subId;
+    }
+    mapping(bytes32 /* requestID */ => Callback) callbacks;
+
+
+    constructor(address _link, address _blockHashStore) public {
+        LINK = LinkTokenInterface(_link);
+        blockHashStore = BlockHashStoreInterface(_blockHashStore);
+    }
+
+    function registerProvingKey(
+        address _oracle, uint256[2] calldata _publicProvingKey
+    )
+    external
+    onlyOwner()
+    {
+        // Must be unique key
+        // I don't think we need the jobID - the serialized PK function as the
+        // identifier of the offchain job to run, since its a strict 1-1 map?
+        // TODO: Save a mapping of keyHash -> {oracle}
+        // Oracle is the address of who gets paid for fulfilling requests.
+    }
+
+    function requestRandomWords(
+        bytes32 keyHash,  // Corresponds to a particular offchain job which uses that key for the proofs
+        uint16  minimumRequestConfirmations,
+        uint16  callbackGasLimit,
+        uint256 subId,   // A data structure for billing
+        uint256 numWords  // Desired number of random words
+    )
+    external
+    returns (bytes32 requestId)
+    {
+       // TODO:
+       // Validate all inputs
+       // Sanity check the subscription has enough link? Just
+       // accept that gas price fluctuations between request and response could potentially
+       // result in request accepted but failed to fulfill.
+       /*
+       nonce = oracles[_keyHash].nonces[msg.sender] + 1;
+       oracles[_keyHash].nonces[msg.sender] = nonce;
+       preSeed = keccak256(abi.encode(keyHash, msg.sender, nonce)); // seed with no blockhash
+       requestId = keccak256(abi.encode(keyHash, preSeed)); // could this just be the preseed to simplify?
```

Comment:
@coventry curious why we couldn't just have preSeed == requestId

Reply:
Can you expand on that approach? The requestID should depend on everything which goes into the preSeed, to prevent replay attacks.

Reply:
sure - I mean ```preSeedAndRequestId = keccak256(abi.encode(keyHash, msg.sender, nonce)```

Reply:
Yeah, I think that's fine.